### DISCUSSION TITLE
Trigger bundle install in C/C++

### DISCUSSION THREAD

**Discussion by dcorbeil** [rauc/rauc#943] [Q&A] [answered]
Hi all,

I'm using a buildroot + rauc + uboot + sysvinit system and want to trigger a rauc install through my C/C++ app. From my understanding, the right way to do so it to do it through D-Bus but I had no success. Documentation and examples are very sparse. I have the rauc service enabled so I can use `dbus-send` to do some dbus operations that way. For example marking a slot as good:

```shell
dbus-send --system --dest=de.pengutronix.rauc --print-reply --type=method_call / de.pengutronix.rauc.Installer.Mark string:"good" string:"rootfs.1"
```

Sadly I can't trigger an update using `dbus-send` since it [doesn't support](https://linux.die.net/man/1/dbus-send) nested containers. e.g.: dict of variants for the install

```
Also, dbus-send does not permit empty containers or nested containers (e.g. arrays of variants).
shell
```

So, here are my questions:
 - Is there an example or documentation for a C/C++ implementation of a D-Bus rauc install?
 - Is there a better way to trigger an install using C/C++ other than D-Bus?

Thank you!

**Comment by jluebbe** [comment, marked as answer]
RAUC itself also uses D-Bus in C to communicate between the CLI and the service (using GDBus/gdbus-codegen from glib). For C++, Qt has good D-Bus integration and code generation. See https://www.freedesktop.org/wiki/Software/DBusBindings/ for many more options.

For the CLI access to D-Bus during development, systemd's busctl much better than dbus-send.

**Reply by dcorbeil** [reply]
Ok thank you. Looking at the rauc code helped a lot.
